### PR TITLE
Focs api

### PR DESCRIPTION
**Pull request by Cjkjvfnby**
This is the first version of the creation of a Python skeleton for the Focs API script.

- It should serve as documentation
-  a source of autocomplete
  изображение
-  and basic type checking 
   изображение 

The current C++ implementation has some caveats
- This API is considered builtin (see the list we added to handle this by linter https://github.com/freeorion/freeorion/blob/master/pyproject.toml#L104). 
- Built in elements are added to the parser in different type, many of them from the beginning, but some of them added when new parsing section is started for example Species.
- prints from the FOCS files are not in logs, but you could see them then start the freeorion.exe, in one of the windows that opened with the game (server?).
изображение

### CODE REVIEW

**Comment by o01eg** [comment]
Current Python FOCS implementation keeps as close as possible to old FOCS script so it uses `globals` to make all parser things a first levels without additional import.

Python FOCS parser use own module loader with root path in `default/scripting` or `test-scripting`.

Objects with `globals` created once per parsed type and then re-used, it doesn't concatenate all FOCS files, it just calls exec on file content with same `globals`.

FreeOrion Logging module isn't initialized for Python FOCS parser.

**Comment by geoffthemedio** [comment]
> * Replace some C++ bindings, for example, if we have > and == defined on the C++ level, we can implement >=, <=, < operations using Python   `a >= 1` could be represented bu with `a == 1 | a > 1` )

That looks like a bad idea, as it would require a multi-part operation internally, rather than a single operation, for the value ref expression tree to represent `>=`.

**Comment by Cjkjvfnby** [comment]
I have found a way to do an API!

I've created an empty `foct.py` when I import an item from it, it will check in this module globals() and return the value. 
Since we manipulate with globals, it always will be there.  Basically, this code does nothing in runtime.  

 As for IDE, I have created `focs.pyi` stub, IDE uses this stub for auto-complete.  (Stub has priority over the original file,  this is used to annotate the libs without changing them.)
 
 Combining this together, I can create focs stubs, without touching any C++ code.

**Review by o01eg — APPROVED** [review]
Looks fine

**Comment by Cjkjvfnby** [comment]
I have updated the skeletons.  

The main change is not we have mypy static analyzer, which will check that code matches the skeleton. This should speed up the development and catch some dummy mistakes. 

The thing that would be useful, is to add all specific objects, at the beginning of the parsing.  This will allow us to avoid these * imports.

File: default/scripting/focs/_species.pyi [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    graphic: str,
+    environments: dict = {},
+    likes: list[str] = [],
+    dislikes: list[str] = [],
+    can_produce_ships: bool = False,
+    playable: bool = False,
+    can_colonize: bool = False,
+    native: bool = False,
+    annexation_condition=None,
+): ...
```
Comment (o01eg) [inline default/scripting/focs/_species.pyi:21]: It always returns `None` object.

File: default/scripting/species/SP_ACIREMA.focs.py [hunk 2]
```diff
@@ -1,6 +1,27 @@
 from common.misc import PLANET_DEFENSE_FACTOR, PLANET_SHIELD_FACTOR
+from focs._effects import (
```
Comment (o01eg) [inline default/scripting/species/SP_ACIREMA.focs.py:2]: Could it be imported with asterisk?

**Review by o01eg — CHANGES_REQUESTED** [review]

File: default/scripting/focs/_species.pyi (hunk 1 above)
Comment (Cjkjvfnby) [reply default/scripting/focs/_species.pyi:21]: No annotation means, that you don't care, as we do in that case.   `...`  is a sign that there is no implementation.

File: default/scripting/species/SP_ACIREMA.focs.py (hunk 2 above)
Comment (Cjkjvfnby) [reply default/scripting/species/SP_ACIREMA.focs.py:2]: Theoretically yes, but I'd prefer to have explicit imports.  If you use modern tools, you won't see this. 

This is how import usually looks like in Pycharm:
изображение

Comment (Cjkjvfnby) [reply default/scripting/species/SP_ACIREMA.focs.py:2]: Also, I put zero effort into managing this section:

I press ALT+ENTER (quick fix) on missed word, and choose fix.  The result is imperfect, but pre-commit with ruff will fix it for me.

изображение

**Comment by geoffthemedio** [comment]
looks like this broke a few test builds...

**Comment by o01eg** [comment]
I'll fix it in #4603

**Pull request merged**